### ISSUE TITLE
Improvement of HybridModel 

### ISSUE THREAD

**Issue by wangcj05** [idaholab/raven#1006]
--------
Issue Description
--------
##### The improvements of current HybridModel include:
1. Able to perform cross validation on time series data;
2. Able to quantify the quality of predictions: : a) estimator score method b) scoring parameter using cross-validation c) metric functions, including classification metrics, multilabel ranking metrics, regression metrics, and clustering metrics
3. Able to preprocess data, such as normalize, standardize, mean removal, variance scaling, centering kernel matrices, non-linear transform, encoding categorical features, custom transfer
4. Able to save the validation/learning data into metadata, and able to use OutStreams to plot it.
5. General strategies to handle under-fitting and overfitting
6. Optional to tune the hyper-parameters of an estimator

##### I think the following methods should be implemented for ROMs:
1. cross validation and cross validation scores;
2. pre-processing data 
3. input data shuffling before training 
4. outlier detection and removal for training 
5. Novelty section for prediction
6. able to handle under-fitting and overfitting of ROMs
7. able to set the minimum training size and maximum training size
8. able to tune hyper-parameters
9. able to check rom convergence based on cross validation and user provided tolerance

##### I think the following methods will be useful for codes:
1.	Should able to provide validation domain, pre-check before run, this will be helpful for hybrid model generalization

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

**Comment by wangcj05** [comment]
Current Input:
```
<Models>
  <ExternalModel ModuleToLoad="EM2linear" name="thermalConductivityComputation" subType="">
    <variables>leftTemperature,rightTemperature,k,averageTemperature</variables>
  </ExternalModel>
  <ROM name="knr" subType="SciKitLearn">
    <SKLtype>neighbors|KNeighborsRegressor</SKLtype>
    <Features>leftTemperature, rightTemperature</Features>
    <Target>k</Target>
    <n_neighbors>5</n_neighbors>
    <weights>uniform</weights>
    <algorithm>auto</algorithm>
    <leaf_size>30</leaf_size>
    <metric>minkowski</metric>
    <p>2</p>
  </ROM>
  <PostProcessor name="pp1" subType="CrossValidation">
      <SciKitLearn>
          <SKLtype>KFold</SKLtype>
          <n_splits>10</n_splits>
          <shuffle>False</shuffle>
      </SciKitLearn>
      <Metric class="Metrics" type="SKL">m1</Metric>
  </PostProcessor>
  <HybridModel name="hybrid" subType="">
      <Model class="Models" type="ExternalModel">thermalConductivityComputation</Model>
      <ROM class="Models" type="ROM">knr</ROM>
      <TargetEvaluation class="DataObjects" type="PointSet">thermalConductivityComputationContainer</TargetEvaluation>
      <CV class="Models" type="PostProcessor">pp1</CV>
      <settings>
          <tolerance>0.005</tolerance>
          <maxTrainSize>1000</maxTrainSize>
          <minInitialTrainSize>10</minInitialTrainSize>
      </settings>
      <validationMethod name="CrowdingDistance">
          <threshold>0.2</threshold>
      </validationMethod>
  </HybridModel>
</Models>
```

Proposed Input:
```
<ROM name="knr" subType="SciKitLearn">

  <SKLtype>neighbors|KNeighborsRegressor</SKLtype>
  <Features>leftTemperature, rightTemperature</Features>
  <Target>k</Target>
  <CV class="Models" type="PostProcessor">pp1</CV>
  <validationMethod name="CrowdingDistance">  <!-- moved to metrics?  -->
      <threshold>0.2</threshold>
  </validationMethod>

  <tolerance>0.005</tolerance>
  <maxTrainSize>1000</maxTrainSize>
  <minInitialTrainSize>10</minInitialTrainSize>

  <settings>
    <n_neighbors>5</n_neighbors>
    <weights>uniform</weights>
    <algorithm>auto</algorithm>
    <leaf_size>30</leaf_size>
    <metric>minkowski</metric>
    <p>2</p>
  </settings>
</ROM>

<HybridModel name="hybrid" subType="">
    <Model class="Models" type="ExternalModel">thermalConductivityComputation</Model>
    <ROM class="Models" type="ROM">knr</ROM>
    <TargetEvaluation class="DataObjects" type="PointSet">thermalConductivityComputationContainer</TargetEvaluation>
    <settings>

    </settings>
</HybridModel>


<Models>
  <HybridModel name="hybrid" subType="">
      <!-- Should be able to handle Code, External Model, ROMs, Ensemble Models? -->
      <Model class="Models" type="Code">codeI</Model>
      <Model class="Models" type="Code">codeII</ROM>
      <TargetEvaluation class="DataObjects" type="PointSet">samples</TargetEvaluation>
      <!-- control functions will used to determine which code to execute -->
      <controlFunctions class="Functions" type="External">control</controlFunctions>
      <!-- <settings>

      </settings> -->
  </HybridModel>
</Models>
```

**Comment by wangcj05** [comment]
@alfoa @PaulTalbot-INL @mandd @joshua-cogliati-inl, I would like to schedule a meeting to discuss the improvement of hybrid model. Please take a look at this issue, and we can have a short discussion at tomorrow stand up meeting. If you think it is worth to schedule a meeting, I will do that after the stand up meeting. Please also feel free to comment on this issue.

Sincerely,

Congjian

**Comment by wangcj05** [comment]
@alfoa Could you provide your feedbacks for this issue?

**Comment by PaulTalbot-INL** [comment]
As a new feature, this should be catalogued for notification at the next release but does not require notification to the user base at this time.

Approved for closure via #1016.

**Comment by wangcj05** [comment]
archived.
